Plugins/Public/kill_tracker/Main.cpp

```diff
@@ -0,0 +1,419 @@
+	{
+		if (i == damageToInflictorMap.rend() || killerCounter >= numberOfKillers)
+		{
+			break;
+		}
+		uint contributionPercentage = static_cast<uint>(round((i->first / totalDamageTaken) * 100));
+		if (contributionPercentage < minimumAssistPercentage)
+		{
+			break;
+		}
+
+		wstring inflictorName = (const wchar_t*)Players.GetActiveCharacterName(i->second);
+		if (killerCounter == 0)
+		{
+			deathMessage = ReplaceStr(deathMessage, L"%killer", inflictorName);
+			deathMessage += L" (" + stows(itos(contributionPercentage)) + L"%)";
+		}
+		else if (killerCounter == 1)
+		{
+			assistMessage = L"Assisted by: " + inflictorName + L" (" + stows(itos(contributionPercentage)) + L"%)";
+		}
+		else
+		{
+			assistMessage += L", " + inflictorName + L" (" + stows(itos(contributionPercentage)) + L"%)";
+		}
+		killerCounter++;
+	}
+
+	deathMessage = L"<TRA data=\"" + killMsgStyle +
+		L"\" mask=\"-1\"/><TEXT>" + XMLText(deathMessage) + L"</TEXT>";
+	if (!assistMessage.empty())
+	{
+		assistMessage = L"<TRA data=\"" + killMsgStyle +
+			L"\" mask=\"-1\"/><TEXT>" + XMLText(assistMessage) + L"</TEXT>";
+	}
+
+	CPlayerGroup* victimGroup = victimGroup = Players[clientVictim].PlayerGroup;
+
+	uint systemId;
+	pub::Player::GetSystem(clientVictim, systemId);
+	Vector victimPos;
+	Matrix victimOri;
+	pub::SpaceObj::GetLocation(Players[clientVictim].iShipID, victimPos, victimOri);
+
+	pd = nullptr;
+	while (pd = Players.traverse_active(pd))
+	{
+		uint playerId = pd->iOnlineID;
+		if (damageArray[playerId][clientVictim] > 0)
+		{
+			HkFMsg(playerId, deathMessage);
+			if(!assistMessage.empty())
+				HkFMsg(playerId, assistMessage);
+			continue;
+		}
+		if (
+		(pd->PlayerGroup &&
+		  ((victimGroup && victimGroup == pd->PlayerGroup)
+		    || ( pd->PlayerGroup && killerGroups.count(pd->PlayerGroup))))
+		|| playerId == clientVictim)
+		{
+			HkFMsg(playerId, deathMessage);
+			if (!assistMessage.empty())
+				HkFMsg(playerId, assistMessage);
+			continue;
+		}
+		if (pd->iSystemID == systemId && Players[playerId].iShipID)
+		{
+			Vector pos;
+			Matrix ori;
+			pub::SpaceObj::GetLocation(Players[playerId].iShipID, pos, ori);
+			if (HkDistance3D(victimPos, pos) <= deathBroadcastRange)
+			{
+				HkFMsg(playerId, deathMessage);
+				if (!assistMessage.empty())
+					HkFMsg(playerId, assistMessage);
+			}
+		}
+	}
+
+	clearDamageTaken(clientVictim);
+}
+
+void __stdcall Disconnect(uint client, enum EFLConnection conn)
+{
+	returncode = DEFAULT_RETURNCODE;
+	uint shipId;
+	pub::Player::GetShip(client, shipId);
+	if (shipId)
+	{
+		PlayerData* pd = nullptr;
+		while (pd = Players.traverse_active(pd))
+		{
+			if (damageArray[pd->iOnlineID][client] != 0)
+			{
+				static wstring victimName = (const wchar_t*)Players.GetActiveCharacterName(client);
+				PrintUserCmdText(pd->iOnlineID, L"%ls : Damaged player %ls has disconnected in space", GetTimeString(false).c_str(), victimName.c_str());
```

Comment:
How long does damage persist? Do you count someone that did damage an hour ago?

Reply:
Damage counters are reset on:
-undock
-character change
-logout

So yes, if someone fights someone, flees and does not dock to repair anywhere for an hour, and the attacker also remains online, it will count.